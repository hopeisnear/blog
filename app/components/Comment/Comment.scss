.comment {
  footer {
    position: relative;
  }
  .author {
    padding-left: 82px;
    .avatar {
      height: 60px;
      width: 60px;
      position: absolute;
      left: 0;
      border-radius: 50%
    }
  }
  .timestamp {
    padding-left: 82px;
  }
  .comment-details {
    margin-top: 20px;
    position: relative;
    background-color: rgba(0,0,0,.05);
    padding: 25px;
    border-radius: 2px;
    p:after {
      border-color: rgba(0,0,0,.05) transparent;
      z-index: 1;
      top: -10px;
      content: '';
      position: absolute;
      border-style: solid;
      border-width: 0 10px 10px;
      display: block;
      width: 0;
      left: 18px;
    }
  }
  .reply-to-comment {
    margin-bottom: 20px;
  }
}
