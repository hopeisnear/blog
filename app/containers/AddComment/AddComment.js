/* eslint-disable jsx-a11y/label-has-for */
/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { PureComponent } from 'react';
import { func, shape, bool } from 'prop-types';
import { VelocityTransitionGroup } from 'velocity-react';
import { FacebookIcon } from 'react-share';

import Login from 'components/LoginFacebook/Login';
import './AddComment.scss';

export default class AddComment extends PureComponent {
  state = {
    showCommentExpandableSection: false,
    content: undefined,
    email: undefined,
    name: undefined,
    website: undefined
  };

  expandCommentDetails = () => {
    this.setState({ showCommentExpandableSection: true });
  };

  handleInputChange = event => {
    this.setState({
      [event.target.name]: event.target.value
    });
  };

  handleSubmit = event => {
    event.preventDefault();

    const { addComment, comment, onAddComment } = this.props;
    const { content, name, email, website } = this.state;
    addComment(comment, { content, name, email, website });
    this.setState({ showCommentExpandableSection: false, content: undefined, email: undefined, name: undefined, website: undefined });
    onAddComment();
  };

  renderLoggedOutPanel() {
    return (
      <React.Fragment>
        <div className="add-comment__login">
          <div>Fill in your details below or click an icon to log in:</div>
          <div className="login__options">
            <Login />
          </div>
        </div>
        <div className="add-comment__expandable-section">
          <div>
            <label className="sticky">
              <input name="email" type="email" size="30" required maxLength="100" value={this.state.email} onChange={this.handleInputChange} />
              <label htmlFor="email">Email</label>
            </label>
          </div>
          <div>
            <label className="sticky">
              <input name="name" type="text" size="30" required maxLength="100" value={this.state.name} onChange={this.handleInputChange} />
              <label htmlFor="name">Name</label>
            </label>
          </div>
          <div>
            <label className="sticky">
              <input name="website" type="url" size="30" maxLength="200" value={this.state.website} onChange={this.handleInputChange} />
              <label htmlFor="website">Website (optional)</label>
            </label>
          </div>
        </div>
      </React.Fragment>
    );
  }

  renderLoggedInPanel() {
    return (
      <div className="add-comment__loggedIn">
        <div className="loggedIn__avatar">
          <img alt="user avatar" width={30} height={30} src={this.props.user.picture} />
        </div>
        <div className="loggedIn__commentingAs">Commenting as</div>
        <div className="loggedIn__username">
          <strong>{this.props.user.name}</strong>
        </div>
        <div className="loggedIn__icon">
          <FacebookIcon size={25} />
        </div>
        <div className="loggedIn__logout">
          <button type="button" className="button secondary small" onClick={this.props.logout}>
            Logout
          </button>
        </div>
      </div>
    );
  }

  renderExpandableSection() {
    return (
      <div>
        {this.props.loggedIn && this.renderLoggedInPanel()}
        {!this.props.loggedIn && this.renderLoggedOutPanel()}
      </div>
    );
  }

  render() {
    const { content, showCommentExpandableSection } = this.state;

    return (
      <div className="AddComment">
        <form onSubmit={this.handleSubmit}>
          <textarea
            className="add-comment__input"
            name="content"
            maxLength="65525"
            required
            placeholder="Enter your comments here"
            value={content}
            onChange={this.handleInputChange}
            onFocus={this.expandCommentDetails}
          />
          <VelocityTransitionGroup enter={{ animation: 'slideDown' }} duration={1500}>
            {showCommentExpandableSection && this.renderExpandableSection()}
            {showCommentExpandableSection && (
              <div className="add-comment__form-submit">
                <input name="submit" type="submit" className="form-submit__input" value="Post Comment" />
              </div>
            )}
          </VelocityTransitionGroup>
        </form>
      </div>
    );
  }
}

AddComment.propTypes = {
  addComment: func.isRequired,
  onAddComment: func,
  comment: shape({}),
  loggedIn: bool,
  user: shape({}),
  logout: func.isRequired
};

AddComment.defaultProps = {
  comment: undefined,
  onAddComment: () => {},
  loggedIn: false,
  user: undefined
};
