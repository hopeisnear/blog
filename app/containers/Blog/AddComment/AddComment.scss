@import "app/styles/variables.scss";

.comment-form-panel {
  max-width: 800px;
  .comment-form-header {
    border: 0;
    margin-bottom: 0;
  }
  .comment-wrapper {
    display: flex;
    flex-direction: column;
    textarea {
      height: 70px;
      background: #F0F0F0;
    }
  }
  .comment-details {
    padding: 0 20px 20px 20px;
    background: #F5F5F5;
    &.hidden {
      display: none;
    }
    .input-wrapper {
      // display: flex;
      // flex-direction: column;
      //margin-bottom: 20px;
    }
  }
  .form-submit {
    padding: 20px;
    display: flex;
    justify-content: flex-end;
    background: #F0F0F0;
    &.hidden {
      display: none;
    }

    input {
      padding: 3px 10px;
      border-radius: 15px;
      background: white;
      font-size: $primary-font-size;
      line-height: $primary-font-line-height;
      color: $primary-color;
      cursor: pointer;
    }
  }
}

label.sticky {
  position: relative;
  > label {
    font-size: 16px;
    margin-left: 1em;
    padding: 0 .2em;
    position: absolute;
    line-height: 1em;
    top: 0;
    left: 0;
    color: #898989;
    transition: all .3s ease-in-out;
    white-space: nowrap;
    background: #fff;
  }
  > input[type="text"], input[type="email"], input[type="url"] {
    box-sizing: border-box;
    font-size: 16px;
    height: 3em;
    width: 100%;
    padding: .5em 1.34em;
    margin-bottom: 20px;
    border: 1px solid #ccc;
    -webkit-font-smoothing: antialiased;
    -webkit-transform: translate(0, 1.34em);
    transition-duration: .1s;
    background: #fff;
    &:focus {
      border: 1px solid #00bafa;
      + label {
        color: #00bafa;

      }
    }
    &:focus, &:valid {
      outline: none;
      + label {
      }
    }
  }
}
