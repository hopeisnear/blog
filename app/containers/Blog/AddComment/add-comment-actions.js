/* eslint-disable no-underscore-dangle */
import { appendComment, addNewComment, addNewReply, appendReply } from 'services/comments-service';
import moment from 'moment';

export const ADD_COMMENT_RESPONDED = 'ADD_COMMENT_RESPONDED';

export function addCommentAction(comment, commentForm) {
  return (dispatch, getState) => {
    const pathname = getState().getIn(['route', 'location', 'pathname'], '/blog/');
    const articleName = pathname.split('/')[2] || '';
    const selectedArticleName = articleName.includes('?') ? articleName.split('?')[0] : articleName;

    const articles = getState().getIn(['blog', 'articles']);
    const selectedArticle = articles.find((article) => article.id === selectedArticleName && article.content !== undefined);

    if (!comment) {
      if (!selectedArticle.comments) {
        return addNewComment(selectedArticle._id, {
          name: commentForm.name,
          content: commentForm.content,
          email: commentForm.email,
          website: commentForm.website,
          createdAt: moment().format()
        }).then((response) => {
          dispatch({ type: ADD_COMMENT_RESPONDED, updatedArticle: response });
        });
      } else {
        return appendComment(selectedArticle._id, {
          name: commentForm.name,
          content: commentForm.content,
          email: commentForm.email,
          website: commentForm.website,
          createdAt: moment().format()
        }).then((response) => {
          dispatch({ type: ADD_COMMENT_RESPONDED, updatedArticle: response });
        });
      }
    } else {
      const commentHaveReply = selectedArticle.comments.some((com) => com._key === comment._key ? com.replies !== undefined : commentHaveReplies(comment._key, com));
      if (!commentHaveReply) {
        return addNewReply(selectedArticle._id, comment._key, {
          name: commentForm.name,
          content: commentForm.content,
          email: commentForm.email,
          website: commentForm.website,
          createdAt: moment().format()
        }).then((response) => {
          dispatch({ type: ADD_COMMENT_RESPONDED, updatedArticle: response });
        });
      } else {
        return appendReply(selectedArticle._id, comment._key, {
          name: commentForm.name,
          content: commentForm.content,
          email: commentForm.email,
          website: commentForm.website,
          createdAt: moment().format()
        }).then((response) => {
          dispatch({ type: ADD_COMMENT_RESPONDED, updatedArticle: response });
        });
      }
    }
  };
}

function commentHaveReplies(searchKey, comment) {
  return comment.replies && comment.replies.some((reply) => {
    if (reply._key === searchKey) {
      return reply.replies !== undefined;
    }
    return commentHaveReplies(searchKey, reply);
  });
}
