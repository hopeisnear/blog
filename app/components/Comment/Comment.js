import React, { PureComponent } from 'react';
import AddComment from 'containers/Blog/AddComment';
import './Comment.scss';

export default class Comment extends PureComponent {
  state = {
    showReplyForm: false
  };

  render() {
    return (
      <article className="comment">
        <footer>
          <div className="author">
            <img alt="" src="https://0.gravatar.com/avatar/f734130d2be486763e433ed0bf4fb6ac?s=60&amp;d=https%3A%2F%2F0.gravatar.com%2Favatar%2Fad516503a11cd5ca435acc9bb6523536%3Fs%3D96" className="avatar" />
            <b className="fn">{this.props.comment.name}</b>
          </div>
          <div className="timestamp">
            <a href="https://blog.fedecarg.com/2017/08/01/website-performance-monitoring-open-source-tool/#comment-43756">
              <time dateTime="2017-08-15T20:55:36+00:00">
                {this.props.comment.createdAt}
              </time>
            </a>
          </div>
        </footer>
        <div className="comment-details">
          <p>
            {this.props.comment.content}
          </p>
        </div>
        <div>
          <button onClick={() => this.setState({ showReplyForm: true })}>reply</button>
          {this.state.showReplyForm && <AddComment comment={this.props.comment} />}
        </div>
      </article>
    );
  }
}
