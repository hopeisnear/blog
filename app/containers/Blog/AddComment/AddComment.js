/* eslint-disable jsx-a11y/label-has-for */
/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { PureComponent } from 'react';
import { func, shape } from 'prop-types';

import './AddComment.scss';

export default class AddComment extends PureComponent {
  state = {
    showCommentExpandableSection: false,
    content: undefined,
    email: undefined,
    name: undefined,
    website: undefined
  };

  expandCommentDetails = () => {
    this.setState({ showCommentExpandableSection: true });
  };

  handleInputChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value
    });
  };

  handleSubmit = (event) => {
    event.preventDefault();

    const { addComment, comment, onAddComment } = this.props;
    const { content, name, email, website } = this.state;
    addComment(comment, { content, name, email, website });
    this.setState({ showCommentExpandableSection: false, content: undefined, email: undefined, name: undefined, website: undefined });
    onAddComment();
  };

  render() {
    const { content, showCommentExpandableSection, email, name, website } = this.state;

    return (
      <div className="AddComment">
        <form onSubmit={this.handleSubmit}>
          <textarea
            className="add-comment__input"
            name="content"
            maxLength="65525"
            required
            placeholder="Enter your comments here"
            value={content}
            onChange={this.handleInputChange}
            onFocus={this.expandCommentDetails}
          />
          {showCommentExpandableSection && (
            <div className="add-comment__expandable-section">
              <div>
                <label className="sticky">
                  <input name="email" type="email" size="30" required maxLength="100" value={email} onChange={this.handleInputChange} />
                  <label htmlFor="email">Email</label>
                </label>
              </div>
              <div>
                <label className="sticky">
                  <input name="name" type="text" size="30" required maxLength="100" value={name} onChange={this.handleInputChange} />
                  <label htmlFor="name">Name</label>
                </label>
              </div>
              <div>
                <label className="sticky">
                  <input name="website" type="url" size="30" maxLength="200" value={website} onChange={this.handleInputChange} />
                  <label htmlFor="website">Website (optional)</label>
                </label>
              </div>
            </div>
          )}
          {showCommentExpandableSection && (
            <div className="add-comment__form-submit">
              <input name="submit" type="submit" className="form-submit__input" value="Post Comment" />
            </div>
          )}
        </form>
      </div>
    );
  }
}

AddComment.propTypes = {
  addComment: func.isRequired,
  onAddComment: func,
  comment: shape({})
};

AddComment.defaultProps = {
  comment: undefined,
  onAddComment: () => {}
};
