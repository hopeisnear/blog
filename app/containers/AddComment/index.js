import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import { addCommentAction, login, logout } from './add-comment-actions';
import AddComment from './AddComment';

const mapDispatchToProps = dispatch => ({
  addComment: (comment, commentForm) => dispatch(addCommentAction(comment, commentForm)),
  login: loginProvider => dispatch(login(loginProvider)),
  logout: () => dispatch(logout())
});

const mapStateToProps = createStructuredSelector({
  loggedIn: state => state.getIn(['login', 'user']) !== undefined,
  user: state => state.getIn(['login', 'user']),
  loginProvider: state => state.getIn(['login', 'loginProvider'])
});

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps
);

export default compose(withConnect)(AddComment);
export { mapDispatchToProps };
