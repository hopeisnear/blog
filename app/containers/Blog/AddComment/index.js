import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import { addCommentAction } from './add-comment-actions';
import AddComment from './AddComment';

const mapDispatchToProps = dispatch => ({
  addComment: (comment, commentForm) => dispatch(addCommentAction(comment, commentForm))
});

const mapStateToProps = createStructuredSelector({
});

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect)(AddComment);
export { mapDispatchToProps };
