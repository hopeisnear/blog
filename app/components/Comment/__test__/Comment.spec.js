import React from 'react';
import { shallow } from 'enzyme';
import AddComment from 'containers/AddComment/Loadable';

import Comment from '../Comment';

describe('<Comment />', () => {
  test('should render <Comment />', () => {
    const commentComponent = shallow(<Comment comment={createComment()} />);

    expect(commentComponent).toMatchSnapshot();
  });

  test('should expand add comment form when reply button is clicked', () => {
    const commentComponent = shallow(<Comment comment={createComment()} />);

    commentComponent.find('button.comment__reply').simulate('click');

    expect(commentComponent.find(AddComment).exists()).toBe(true);
  });

  function createComment() {
    return { name: 'John Smith', createdAt: '2018-08-15T18:12:36+02:00', content: 'this is some test comment' };
  }
});
