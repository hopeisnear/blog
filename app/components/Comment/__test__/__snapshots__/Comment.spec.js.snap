// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Comment /> should render <Comment /> 1`] = `
<article
  className="Comment"
>
  <footer
    className="comment__footer"
  >
    <div
      className="comment__author"
    >
      <img
        alt="avatar"
        className="author__avatar"
        src="https://0.gravatar.com/avatar/f734130d2be486763e433ed0bf4fb6ac?s=60&d=https%3A%2F%2F0.gravatar.com%2Favatar%2Fad516503a11cd5ca435acc9bb6523536%3Fs%3D96"
      />
      <b>
        John Smith
      </b>
    </div>
    <div
      className="comment__timestamp"
    >
      <a
        disabled={true}
        href="http://TODO/#comment-43756"
      >
        <time
          dateTime="2018-08-15T18:12:36+02:00"
        >
          August 15, 2018
        </time>
      </a>
    </div>
  </footer>
  <div
    className="comment__body"
  >
    <p
      className="comment__content"
    >
      this is some test comment
    </p>
  </div>
  <div>
    <button
      className="button comment__reply"
      onClick={[Function]}
      type="button"
    >
      reply
    </button>
  </div>
</article>
`;
